pkg/controllers/deploymentreplicassyncer/deployment_replicas_syncer_controller_test.go

```diff
@@ -0,0 +1,585 @@
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	policyv1alpha1 "github.com/karmada-io/karmada/pkg/apis/policy/v1alpha1"
+	workv1alpha2 "github.com/karmada-io/karmada/pkg/apis/work/v1alpha2"
+	"github.com/karmada-io/karmada/pkg/util"
+	"github.com/karmada-io/karmada/pkg/util/names"
+)
+
+func TestPredicateFunc(t *testing.T) {
+	testCases := []struct {
+		name     string
+		oldObj   *appsv1.Deployment
+		newObj   *appsv1.Deployment
+		expected bool
+	}{
+		{
+			name: "retain-replicas label added",
+			oldObj: &appsv1.Deployment{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{},
+				},
+			},
+			newObj: &appsv1.Deployment{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{util.RetainReplicasLabel: util.RetainReplicasValue},
+				},
+			},
+			expected: true,
+		},
+		{
+			name: "replicas changed",
+			oldObj: &appsv1.Deployment{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{util.RetainReplicasLabel: util.RetainReplicasValue},
+				},
+				Spec: appsv1.DeploymentSpec{
+					Replicas: ptr.To[int32](3),
+				},
+				Status: appsv1.DeploymentStatus{
+					Replicas: 3,
+				},
+			},
+			newObj: &appsv1.Deployment{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{util.RetainReplicasLabel: util.RetainReplicasValue},
+				},
+				Spec: appsv1.DeploymentSpec{
+					Replicas: ptr.To[int32](3),
+				},
+				Status: appsv1.DeploymentStatus{
+					Replicas: 4,
+				},
+			},
+			expected: true,
+		},
+		{
+			name: "no relevant changes",
+			oldObj: &appsv1.Deployment{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{util.RetainReplicasLabel: util.RetainReplicasValue},
+				},
+				Spec: appsv1.DeploymentSpec{
+					Replicas: ptr.To[int32](3),
+				},
+				Status: appsv1.DeploymentStatus{
+					Replicas: 3,
+				},
+			},
+			newObj: &appsv1.Deployment{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{util.RetainReplicasLabel: util.RetainReplicasValue},
+				},
+				Spec: appsv1.DeploymentSpec{
+					Replicas: ptr.To[int32](3),
+				},
+				Status: appsv1.DeploymentStatus{
+					Replicas: 3,
+				},
+			},
+			expected: false,
+		},
+		{
+			name:     "create event",
+			oldObj:   nil,
+			newObj:   &appsv1.Deployment{},
+			expected: false,
+		},
+		{
+			name:     "delete event",
+			oldObj:   &appsv1.Deployment{},
+			newObj:   nil,
+			expected: false,
+		},
+		{
+			name:     "generic event",
+			oldObj:   nil,
+			newObj:   &appsv1.Deployment{},
+			expected: false,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			var result bool
+			switch {
+			case tc.name == "create event":
+				result = predicateFunc.Create(event.CreateEvent{Object: tc.newObj})
+			case tc.name == "delete event":
+				result = predicateFunc.Delete(event.DeleteEvent{Object: tc.oldObj})
+			case tc.name == "generic event":
```

Comment:
Generally, the name is read-only and may be changed frequently in the future. If the switch case statement is used, it is recommended that a type field be added to indicate the name.

Reply:
I'll add an event type structure.